### ISSUE TITLE
Get Next Job id?

### ISSUE THREAD

**Issue by Siecje** [kr/beanstalkd#379]
Is there a way to get the next job id for a tube? And once it is given it won't be used for the auto generated id. To use the job id you give it when you add a job.

I am tracking jobs in a database so I can add the job to the database before adding the beanstalk job and then update it.

```
Add to database without beanstalk id
Add to beanstalkd
# if it fails here there will be a beanstalk job that is not tracked
Update database with beanstalkd id
```

On recovery if there is a database entry without a beanstalk job it will add a beanstalk job. The beanstalk job will be added twice.

If I add the beanstalk job first it won't be able to lookup the job in the database.

**Comment by kr** [comment]
That's a good idea. Beanstalkd doesn't currently support this, but it would be nice to add.

**Comment by JensRantil** [comment]
Since it would be unfeasible to store all "reserved ids" in beanstalkd this proposal would mean that a job id can be reused. That would break backwards compatibility and sounds a little scary to me.

@Siecje Just curious, what's the reason for you tracking the jobs in a database?

**Comment by wjrogers** [comment]
I'm doing something similar, so I might be able to offer a little help. I track jobs in a database for a) easier monitoring and b) dependency logic (e.g. don't dispatch job 2 until job 1 completes). I do the same thing as you:

```
Add to database without beanstalk id
Add to beanstalkd
# if it fails here there will be a beanstalk job that is not tracked
Update database with beanstalkd id
```

I mitigate the untracked beanstalk job problem by assigning each job a guid. I change the status of the job from Queued to Dispatched in your step 3 (when storing the beanstalk ID in the database). When a worker reserves a job, I look the job up in the database. If the job is still Queued, the worker has outraced the producer, so it releases the job. If the job doesn't exist in the database or if the beanstalk ID in the database doesn't match the one the worker reserved, the worker deletes the job. This works well for me.

**Comment by JensRantil** [comment]
Good input. Thanks, @wjrogers. Your comment reminded me of https://github.com/beanstalkd/beanstalkd/pull/180#pullrequestreview-147460651.

**Comment by Siecje** [comment]
@JensRantil You don't need to store all "reserved ids".

When you add a job to a tube you get an id. I just want to say "Hey beanstalkd what is the next id you are going to give out. Thanks, don't give that id out. Here is my job and this is the job id."

In the mean time if jobs are added they will get a different id.

**Comment by ysmolsky** [comment]
This might a good feature, but I see the problem with breaking promises of the protocol. We need to modify the `put` command. While the reserving command could be added without problem.

Also the answers above showed how it is possible to handle your case on the side of a client. Could that work out for you?

**Comment by Siecje** [comment]
Create a new command? 

I have not seen a solution for this problem.

**Comment by kr** [comment]
Yeah, I think this would take 2 new commands, one to reserve the ID, and another to put a job with the reserved ID. Something like `reserve-id` and `put-with-id`. The existing `put` command would have to keep its existing behavior of reserving and immediately using a new ID.

If done this way, it would not be a breaking change.

There is an open question about what to do if the client tries to use an ID that wasn't previously reserved, or if the client uses a reserved ID more than once. I think it would be wise to reject those cases; this would necessitate storing a list of reserved IDs that haven't been used yet (and writing it to the WAL and restoring it during recovery — this would involve some design work to extend the WAL data format for a new pair of record types).

Note that the proposed `put-with-id` command could be idempotent, while the existing `put` command can not.

**Comment by ysmolsky** [comment]
This features is somewhat helps to solve some of usecases of #2. Client can pick the the id before adding the job and can store the job with specific id too. The main issue with reserving ID mechanism. Should we allow to pick any ID or should we ask for next ID to be returned by beanstalkd and then use it? What happens if client reserves some ID and does not use it?

**Comment by kr** [comment]
> Should we allow to pick any ID or should we ask for next ID to be returned by beanstalkd and then use it?

I would suggest letting beanstalkd pick the actual ID value, like it already does.

The command would be:

```
reserve-id\r\n
```

With a response of:

```
ID <id>\r\n
```

And no error cases.

> What happens if client reserves some ID and does not use it?

That's a great question. I think the simplest answer is that the ID remains unused and reserved forever. But then it would be possible to accumulate an unbounded amount of reserved IDs over time. So there ought to be a way to clean them up.

One way might be to associate a reserved ID to the connection that reserved it, sort of like how we associate a reserved job to the connection that reserved it. If the connection is closed, then any reserved IDs it had are discarded, never to be used.

This would also let us avoid writing anything or making any changes to the WAL. When the server restarts, all connections are closed anyway, so any IDs that had been reserved ought to be discarded.

**Comment by ysmolsky** [comment]
> One way might be to associate a reserved ID to the connection that reserved it, sort of like how we associate a reserved job to the connection that reserved it. If the connection is closed, then any reserved IDs it had are discarded, never to be used.

There might be a slight problem with this approach. What if beanstalkd restarts after the client has issued "reserve-id" and before the "put-with-id"? The obvious answer to throw an error is not good, client has just wasted the time to upload the job. This case has to be handled on the client, whenever disconnect happens, all bets are off and client has to reissue "reserve-id".

**Comment by kr** [comment]
> The obvious answer to throw an error is not good, client has just wasted the time to upload the job. This case has to be handled on the client, whenever disconnect happens, all bets are off and client has to reissue "reserve-id".

Are disconnections (including ones that happen because the server restarted) frequent enough for this to be a problem?

**Comment by kr** [comment]
But yes, you’re right. If the server associates reserved IDs with a connection, the client should do that too. If it loses a connection, all the reservations on that connection would become invalid, so the client should throw them away.

**Comment by onyxbb** [comment]
I think an optional ID argument for the put command or put-with-id is terrific because if we can post the same job to two or more beanstalkd instances then redundancy/fail-over becomes possible. Consumers can select from a pool of hosts and once a job is retrieved use the ID to update the job status on the other instances. With an async client it shouldn't slow things down too much neither.

**Comment by kr** [comment]
> if we can post the same job to two or more beanstalkd instances then redundancy/fail-over becomes possible. Consumers can select from a pool of hosts and once a job is retrieved use the ID to update the job status on the other instances.

I agree this sort of ability would be nice. But in the proposed design here, it is not possible to guarantee that the same ID can be used for the same job in two different beanstalkd instances.

If we want that guarantee, it would require behavior more like #2. The fundamental difference is that this issue uses beanstalkd-supplied values for the job ID, meaning the client cannot control which values are available, whereas issue #2 uses client-supplied values for the job name. (The design in #2 says that a string value would be used for the name, but it could just as easily be a number. That detail is not particularly relevant here.)

That leaves the two existing benefits of this proposal:

- The job ID can be saved in a database before the job is submitted to beanstalkd.
- Regardless of how many beanstalkd instances are involved, this feature would let the client ensure submitting a job is idempotent.

I'm not sure which approach is better. Issue #2 is more difficult to implement, but it also has the additional benefit you've identified.

**Comment by onyxbb** [comment]
Read previous discussion https://github.com/beanstalkd/beanstalkd/issues/264 on this subject. In response to some of @JensRantil concerns:

> Allowing a custom ID also has the implications that we break the uniqueness of job id. There are also potential concurrency confusions since a very recently deleted job might seem to "pop up" again if a job with a specific job id is added adding lots of confusion to both clients and developers.

Key uniqueness could be ensured by either affixing or suffixing the client supplied key to the one generated by beanstalkd as per usual. e.g.

beanstalkd generated job ID: 2828
client supplied job ID: 7744
resulting key ID = 2828.7744

For this I imagine key ID type has to be changed from unsigned int64 to an unsigned type that supports decimals (floats? sorry not a C developer).

If no job ID is supplied from the client, then beanstalkd just uses its own job ID (backwards compatibility?).

Although key ID will no longer be unique across different beanstalkd instances, clients can take the decimals as the client supplied ID to match across beanstalkd instances.

But IMO just allowing the client to submit the key ID makes most sense unless a re-using a previously used job ID will cause problems somehow. In that case I propose the above workaround.

**Comment by chekiboy** [comment]
Hi, as I see this feature has been merged into master. (reserve-job) but there is no new version from that. Could you create a new version?

**Comment by tomponline** [comment]
@chekiboy we plan to make a new release soon. Thanks

**Comment by chekiboy** [comment]
> @chekiboy we plan to make a new release soon. Thanks

Good to hear it. Thanks for informing me :)

**Comment by chekiboy** [comment]
@tomponline Could you give me an update about this thread? This is a required dependency on our new development. :/

**Comment by tomponline** [comment]
@chekiboy release v1.12 has been done https://github.com/beanstalkd/beanstalkd/releases/tag/v1.12

**Comment by tomponline** [comment]
@ysmolsky can we close this now?

**Comment by ysmolsky** [comment]
I think there is some confusion about this. I am sorry not to answer sooner. We never implemented commands mentioned in this issue. We merely added "reserve-job" that reserves an existing job by the id. 

This issue is yet to be decided upon, not speaking about it's implementation.

**Comment by JiaRG** [comment]
Any progress on this feature?
